iroh-net/src/endpoint.rs

```diff
@@ -109,12 +109,18 @@ impl Builder {
 
     /// Binds the magic endpoint on the specified socket address.
     ///
-    /// The *bind_port* is the port that should be bound locally.
-    /// The port will be used to bind an IPv4 and, if supported, and IPv6 socket.
-    /// You can pass `0` to let the operating system choose a free port for you.
+    /// The *addr_v4* is the address that should be bound locally on IPv4.
     ///
-    /// NOTE: This will be improved soon to add support for binding on specific addresses.
-    pub async fn bind(self, bind_port: u16) -> Result<Endpoint> {
+    /// You can pass `None` to let the operating system choose a free port for you, and bind to `0.0.0.0`.
+    ///
+    /// The *addr_v6* is the address that should be bound locally on IPv6.
+    ///
+    /// You can pass `None` to let the operating system choose a free port for you, and bind to `::1`.
+    pub async fn bind(
+        self,
+        addr_v4: Option<SocketAddrV4>,
+        addr_v6: Option<SocketAddrV6>,
+    ) -> Result<Endpoint> {
```

Comment:
Maybe an `Option` here is a bit odd?  E.g. I can't request to only bind on IPv4 right now because `None` does mean "os chooses" which is a bit surprising.

We also have the slightly odd thing that we will iterate over ports until one is selected, we should at least document that for now.

I'm wondering if a nicer API could be to add new builder methods and make the bind call not take any argument:
```
builder
    .bind_ipv4(Option<SocketAddrV4>)
    .bind_ipv6(Option<SocketAddrV6>)
    .build()
```
This way `None` can mean, don't bind (though we don't currently support not binding IPv4, I think we can just error for now and leave that for the future).  And the default can be the current default but without having to write it out in full.

I would be tempted to also fail if we can't bind the specific port requested if one is used here.  The only question is then how to handle the way the `iroh::node::Node` currently binds: it requests a specific port but is happy to fall back to `0` if it is not available.  Maybe a `.fallback_port(u16)` to the builder?  Not great.  Maybe allow passing in bound sockets directly?  But that's not great as we need a specifically set-up socket.